hack/build-and-push.sh

```diff
@@ -133,95 +180,324 @@ if [ "$SKIP_BUILD" == "" ]; then
-        if [[ "${task_description}" != "null" ]]; then
-            ANNOTATIONS+=("org.opencontainers.image.description=${task_description}")
-        fi
-        if [ -f "${task_dir}/README.md" ]; then
-            ANNOTATIONS+=("org.opencontainers.image.documentation=${VCS_URL}/tree/${VCS_REF}/${task_dir}README.md")
+# Determine if a task is a kustomized task. 0 returns if it is, otherwise 1 is returned.
+is_kustomized_task() {
+    local -r task_dir=$1
+    local -r task_name=$2
+    local -r kt_config_file="$task_dir/kustomization.yaml"
+    local -r task_file="${task_dir}/${task_name}.yaml"
+    if [ -f "$kt_config_file" ] && [ ! -e "$task_file" ]; then
+        return 0
+    fi
+    return 1
+}
+
+# Generates task bundle with tag. The result bundle reference can be configured
+# by environment variable TEST_REPO_NAME for testing purpose.
+# Arguments: task_name, task_version
+# Task bundle reference is output to stdout.
+generate_tagged_task_bundle() {
+    local -r task_name=$1 task_version=$2
+    local -r repository=${TEST_REPO_NAME:-task-${task_name}}
+    local -r tag=${TEST_REPO_NAME:+${task_name}-}${task_version}
+    echo "quay.io/${QUAY_NAMESPACE}/${repository}:${tag}"
+}
+
+# Generate build data for a normal task. The data is output to stdout as
+# space-separated fields in a single line.
+# Arguments: task_dir, task_name, task_version
+generate_normal_task_build_data() {
+    local -r task_dir=$1 task_name=$2 task_version=$3
+    local -r task_file="${task_dir}/${task_name}.yaml"
+    local -r prepared_task_file="${WORKDIR}/${task_name}-${task_version}.yaml"
+    cp "$task_file" "$prepared_task_file"
+    local -r task_file_sha=$(git log -n 1 --pretty=format:%H -- "$task_file")
+    local -r task_bundle=$(generate_tagged_task_bundle "$task_name" "$task_version")
+    echo "$prepared_task_file $task_file_sha $task_bundle"
+}
+
+# Generate build data for a normal task. The data is output to stdout as
+# space-separated fields in a single line.
+# Arguments: task_dir, task_name, task_version
+generate_kustomized_task_build_data() {
+    local -r task_dir=$1 task_name=$2 task_version=$3
+    local -r task_file="$task_dir/$task_name.yaml"
+    local -r prepared_task_file="${WORKDIR}/${task_name}-${task_version}.yaml"
+    oc kustomize "$task_dir" >"$prepared_task_file"
+    local -r task_file_sha=$(sha256sum "$prepared_task_file" | awk '{print $1}')
+    local -r task_bundle=$(generate_tagged_task_bundle "$task_name" "$task_version")
+    echo "$prepared_task_file $task_file_sha $task_bundle"
+}
+
+declare -r ENABLE_CACHE=${ENABLE_CACHE:-""}
+CACHE_FILE=
+
+if [ -n "$ENABLE_CACHE" ]; then
+    CACHE_FILE="/tmp/build-definitions-build-and-push.cache"
+    declare -r CACHE_FILE
+fi
+
+cache_get() {
+    local -r key=$1
+    local value=
+    if [ -n "$ENABLE_CACHE" ]; then
+        # No match should not fail the script. It means no cached data.
+        read -r _ value < <(grep "^${key}" "$CACHE_FILE" 2>/dev/null) || :
```

Comment:
Would this work? The `grep` seems potentially unreliable
```bash
value=$(awk -v key="$key" '$1 == key { print $2 }' < "$CACHE_FILE")
```

Or, to find the last occurrence of the key (since `cache_set` just appends keys):

```bash
value=$(awk -v key="$key" '$1 == key { value = $2 } END { print value }' < "$CACHE_FILE")

```